src/main/scala/workflow/RuleExecutor.scala

```diff
@@ -0,0 +1,85 @@
+package workflow
+
+import pipelines.Logging
+
+abstract class RuleExecutor extends Logging {
```

Comment:
The RuleExecutor - we're lifting from Catalyst, right? We can't use the Catalyst one directly because it's marked private or some other reason? (I really hate copy-pasting code!)